core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -805,13 +814,12 @@ private[spark] class BlockManager(
           }
         }
       } finally {
-        // If we failed in putting the block to memory/disk, notify other possible readers
-        // that it has failed, and then remove it from the block info map.
-        if (!marked) {
-          // Note that the remove must happen before markFailure otherwise another thread
-          // could've inserted a new BlockInfo before we remove it.
-          blockInfo.remove(blockId)
-          putBlockInfo.markFailure()
+        if (!blockWasSuccessfullyStored) {
+          // Guard against the fact that MemoryStore might have already removed the block if the
+          // put() failed and the block could not be dropped to disk.
+          if (blockInfoManager.lockForWriting(blockId, blocking = false).isDefined) {
```

Comment:
Ohhhh, I think I got it. Let's say you try to put a block `MEMORY_ONLY` but there is not enough space, in which case we will call `blockManager.dropFromMemory` on that block. This call also acquires the write lock and releases the lock in the end. However, the way this whole locking thing works is that if I acquire the write lock twice, a single call to `unlock` will release both locks, so if that happens we need to acquire the write lock again here in this line.